```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

```

## Plotting Genomic Markers

The function **kpPlotMarkers** can be used to plot markers along the genome,
that is, specific positions of the genome with a name: genes, snps, etc...




```{r Figure1, message=FALSE, warning=FALSE}

library(karyoploteR)

kp <- plotKaryotype()

```

It is possible to specify other genomes. If it's one of the few included in the 
package it will be plotted from cache and if it's not, karyoploteR will try to load the 
corresponding BSGenome library and connect to UCSC to download the the cytobands. The 
BSGenome needs to be installed in the system. Use **installed.genomes()** from package 
**BSGenome** to list them.

```{r Figure2, message=FALSE, warning=FALSE}

#Drosophila
kp <- plotKaryotype(genome = "dm6")

```



